src/lib.rs

```diff
@@ -31,17 +43,36 @@ fn relaunch() -> Result<Child, Error> {
 }
 
 fn pid1_handling(child: Option<Child>) -> ! {
-    let child = child.map(|x| x.id());
     let mut signals = Signals::new([SIGTERM, SIGINT, SIGCHLD]).unwrap();
+
+    let child = child.map(|x| x.id());
+    struct ExitStatus {
+        pid: Pid,
+        exit_code: i32,
+    }
+
     loop {
-        for signal in signals.pending() {
+        for signal in signals.forever() {
             if signal == SIGTERM || signal == SIGINT {
+                // TODO: Should forward these signals to the
+                // application and then force kill the application
+                // after certain time.
                 graceful_shutdown();
             }
             if signal == SIGCHLD {
                 let pid = match nix::sys::wait::wait().unwrap() {
-                    WaitStatus::Exited(pid, _) => Some(pid),
-                    WaitStatus::Signaled(pid, _, _) => Some(pid),
+                    WaitStatus::Exited(pid, exit_code) => {
+                        let exit_status = ExitStatus { pid, exit_code };
+                        Some(exit_status)
+                    }
+                    WaitStatus::Signaled(pid, signal, _) => {
+                        let exit_status = ExitStatus {
+                            pid,
+                            // Translate signal to exit code
+                            exit_code: signal as i32 + 128,
```

Comment:
Why are we sometimes returning the pid and sometimes the exit code/signal?

Reply:
The process can exit because of two reasons:

- Normal exit
- Exit via Unix signals

In case of Unix signals, we remap it to proper exit code.

Reply:
I get that, but you seem to be conflating the process ID and the exit status.

Reply:
Do you think renaming the struct `ExitStatus` to `ProcessStatus` would be helpful ?

Reply:
Do you think renaming it from `ExitStatus` to `ProcessStatus` will be helpful ?

Reply:
Can you look over the code one more time from scratch? I think you'll see what I'm talking about: in some branches you're returning a process ID, in some an exit code, and in some a signal, but assigning all of them to a variable called pid.

Reply:
I have gone ahead with some renaming which will hopefully help with the confusion.

Reply:
> Can you look over the code one more time from scratch? I think you'll see what I'm talking about: in some branches you're returning a process ID, in some an exit code, and in some a signal, but assigning all of them to a variable called pid.

Sure, I will let you know once it's ready for review. Thanks!

Reply:
> Can you look over the code one more time from scratch? I think you'll see what I'm talking about: in some branches you're returning a process ID, in some an exit code, and in some a signal, but assigning all of them to a variable called pid.

Sure, I will let you know once it's ready for review. Thanks!

Reply:
Never mind, I see my confusion, you were completely correct. Sorry for the noise.